src/main/java/ymfc/parser/Parser.java

```diff
@@ -239,11 +241,28 @@ private static AddCommand getAddRecipeCommand(String args) throws InvalidArgumen
         String trimmedName = name.trim();
         String ingredString = m.group("ingreds");
         String stepString = m.group("steps");
+
+        // Collect ingredients
+        List<String> ingredientList = Arrays.stream(ingredString.split("\\s+[iI]/"))
+                .map(String::trim)
+                .filter(s -> !s.isEmpty())
+                .toList();
+
+        // Check for duplicates
+        HashSet<String> ingredientSet = new HashSet<>();
+        for (String ingredient : ingredientList) {
+            if (!ingredientSet.add(ingredient.toLowerCase())) {
+                throw new InvalidArgumentException("Duplicate ingredient found: " + ingredient
+                        + ".\nThis is not allowed!");
+            }
+        }
+
         ArrayList<Ingredient> ingreds = Arrays.stream(ingredString.split("\\s+[iI]/"))
                 .map(String::trim)
                 .filter(s -> !s.isEmpty())
                 .map(Ingredient::new)
                 .collect(Collectors.toCollection(ArrayList::new));
```

Comment:
I think these part could be somehow merged together, but this is good for now as we are running out of time. Great work! 👍